File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/document_entries/IDocumentEntry.java
package com.jasonzqshen.familyaccounting.core.document_entries;


import com.jasonzqshen.familyaccounting.core.exception.MandatoryFieldIsMissing;
import com.jasonzqshen.familyaccounting.core.exception.NoFieldNameException;
import com.jasonzqshen.familyaccounting.core.exception.NotInValueRangeException;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;

public interface IDocumentEntry {

	void setValue(String fieldName, Object value) throws NoFieldNameException,
			NotInValueRangeException;

	Object getValue(String fieldName) throws NoFieldNameException;

	Object getDefaultValue(String fieldName) throws NoFieldNameException;
	
	Object[] getValueSet(String fieldName) throws NoFieldNameException;

	/**
	 * check before save
	 * 
	 * @throws MandatoryFieldIsMissing
	 */
	void checkBeforeSave() throws MandatoryFieldIsMissing;

	/**
	 * save document
	 * 
	 * @throws MandatoryFieldIsMissing
	 */
	void save(boolean saved) throws MandatoryFieldIsMissing;

	/**
	 * is saved
	 * 
	 * @return
	 */
	boolean isSaved();

	/**
	 * get document, after save
	 * 
	 * @return
	 */
	HeadEntity getDocument();
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/DocumentEntries/IDocumentEntry.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;

namespace WinStore_FamilyAccountingCore.DocumentEntries
{
    public interface IDocumentEntry
    {
        /// <summary>
        /// set value
        /// </summary>
        /// <param name="fieldName"></param>
        /// <param name="value"></param>
        /// <exception cref="NoFieldNameException">No such field name</exception>
        /// <exception cref="NotInValueRangeException">The value is not supported</exception>
        void SetValue(String fieldName, Object value);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="fieldName"></param>
        /// <returns></returns>
        /// <exception cref="NoFieldNameException"> no such field name</exception>
        Object GetValue(String fieldName);

        /// <summary>
        /// 
        /// </summary>
        /// <param name="fieldName"></param>
        /// <returns></returns>
        /// <exception cref="NoFieldNameException"> no such field name</exception>
        Object GetDefaultValue(String fieldName)
            ;

        /// <summary>
        /// 
        /// </summary>
        /// <param name="fieldName"></param>
        /// <returns></returns>
        /// <exception cref="NoFieldNameException"> no such field name</exception>
        List<MasterDataBase> GetValueSet(String fieldName);

        /// <summary>
        /// Check the document entry before saving
        /// </summary>
        /// <exception cref="MandatoryFieldIsMissing"></exception>
        void CheckBeforeSave();

        /// <summary>
        /// 
        /// </summary>
        /// <param name="saved"></param>
        /// <exception cref="MandatoryFieldIsMissing"></exception>
        /// <exception cref="SaveClosedLedgerException"></exception>
        Task SaveAsync(bool saved);

        /// <summary>
        /// is saved
        /// </summary>
        /// <returns></returns>
        bool isSaved();

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        HeadEntity getDocument();
    }

}
